tools/Dashboard/DevHome.Dashboard/Services/WidgetAdaptiveCardRenderingService.cs

```diff
@@ -0,0 +1,139 @@
+using DevHome.Dashboard.Helpers;
+using Microsoft.UI.Xaml;
+using Windows.Storage;
+using WinUIEx;
+
+namespace DevHome.Dashboard.Services;
+
+public class WidgetAdaptiveCardRenderingService : IAdaptiveCardRenderingService, IDisposable
+{
+    public event EventHandler RendererUpdated = (_, _) => { };
+
+    private readonly WindowEx _windowEx;
+
+    private readonly IThemeSelectorService _themeSelectorService;
+
+    private readonly SemaphoreSlim _rendererLock = new(1, 1);
+
+    private AdaptiveCardRenderer _renderer;
+
+    private bool _disposedValue;
+
+    public WidgetAdaptiveCardRenderingService(WindowEx windowEx, IThemeSelectorService themeSelectorService)
+    {
+        _windowEx = windowEx;
+        _themeSelectorService = themeSelectorService;
+        _themeSelectorService.ThemeChanged += OnThemeChanged;
+    }
+
+    public void Dispose()
+    {
+        Dispose(disposing: true);
+        GC.SuppressFinalize(this);
+    }
+
+    protected virtual void Dispose(bool disposing)
+    {
+        if (!_disposedValue)
+        {
+            if (disposing)
+            {
+                _rendererLock.Dispose();
+            }
+
+            _disposedValue = true;
+        }
+    }
+
+    public async Task<AdaptiveCardRenderer> GetRenderer()
+    {
+        // We need to lock the renderer, otherwise another widget could come in after the renderer
+        // is created but before it is configured and render the widget without configuration.
+        await _rendererLock.WaitAsync();
+        try
+        {
+            if (_renderer == null)
+            {
+                _renderer = new AdaptiveCardRenderer();
+                await ConfigureWidgetRenderer();
+            }
+
+            return _renderer;
+        }
+        finally
+        {
+            _rendererLock.Release();
+        }
+    }
+
+    private async Task ConfigureWidgetRenderer()
+    {
+        // Add custom Adaptive Card renderer.
+        _renderer.ElementRenderers.Set(LabelGroup.CustomTypeString, new LabelGroupRenderer());
+        _renderer.ElementRenderers.Set("Input.ChoiceSet", new AccessibleChoiceSet());
+
+        // A different host config is used to render widgets (adaptive cards) in light and dark themes.
+        await UpdateHostConfig();
+
+        // Due to a bug in the Adaptive Card renderer (https://github.com/microsoft/AdaptiveCards/issues/8840)
+        // positive and destructive buttons render with square corners. Override with XAML styles.
+        var positiveStyle = Application.Current.Resources["AccentButtonStyle"] as Style;
+        var destructiveStyle = Application.Current.Resources["DefaultButtonStyle"] as Style;
+
+        _renderer.OverrideStyles = new ResourceDictionary
+        {
+            { "Adaptive.Action.Positive", positiveStyle },
+            { "Adaptive.Action.Destructive", destructiveStyle },
+        };
+    }
+
+    private async Task UpdateHostConfig()
+    {
+        if (_renderer != null)

```

Comment:
Can `_renderer` become `null` from another thread?

Reply:
There's no way to un-set _renderer. IIRC, this is protecting against the case where it's called by OnThemeChanged because someone subscribed to those events, but hasn't actually used the renderer and therefore created it yet. (Contrary to what GitHub thinks, this file is renamed here, not added, hence needing to remember.)